src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -349,44 +347,46 @@ public static unsafe bool TryFormatDecimal(decimal value, ReadOnlySpan<char> for
             return sb.TryCopyTo(destination, out charsWritten);
         }
 
-        internal static unsafe void DecimalToNumber(ref decimal d, ref NumberBuffer number)
+        internal static void DecimalToNumber(ref decimal d, ref NumberBuffer number)
         {
-            byte* buffer = number.GetDigitsPointer();
+            Span<byte> buffer = number.Digits;
             number.DigitsCount = DecimalPrecision;
             number.IsNegative = d.IsNegative;
 
-            byte* p = buffer + DecimalPrecision;
+            int p = DecimalPrecision;
             while ((d.Mid | d.High) != 0)
             {
-                p = UInt32ToDecChars(p, decimal.DecDivMod1E9(ref d), 9);
+                p = UInt32ToDecChars(buffer.Slice(0, p), decimal.DecDivMod1E9(ref d), 9);
             }
-            p = UInt32ToDecChars(p, d.Low, 0);
+            p = UInt32ToDecChars(buffer.Slice(0, p), d.Low, 0);
 
-            int i = (int)((buffer + DecimalPrecision) - p);
+            int i = DecimalPrecision - p;
 
             number.DigitsCount = i;
             number.Scale = i - d.Scale;
 
-            byte* dst = number.GetDigitsPointer();
+            int dst = 0;
             while (--i >= 0)
             {
-                *dst++ = *p++;
+                buffer[dst++] = buffer[p++];
```

Comment:
I wonder if `Span.CopyTo` would be better here.
It's only up to 29 digits, so this could (realistically) be a couple SIMD loads/stores (with overlap).

Not necessarily important to test/validate, but was just a thought.